### PR TITLE
Convert wealth_dynamics lecture from Numba to JAX

### PR DESCRIPTION
**Pull request by mmcky**
## Summary

This PR converts the wealth_dynamics lecture from Numba to JAX following the QuantEcon style guide rules.

### JAX Conversion (following style guide rules):
- **qe-jax-002**: Added JAX GPU admonition warning about hardware acceleration
- **qe-jax-005, qe-jax-006**: Eliminated @jitclass, converted WealthDynamics to NamedTuple + factory function pattern
- **qe-jax-003, qe-jax-007**: Converted to pure functions with no mutation, added JAX transformations (@jax.jit, jax.vmap)
- **qe-jax-009**: Implemented explicit JAX random key management throughout
- **qe-jax-008**: Replaced explicit loops with jax.lax.scan for optimal performance
- **qe-jax-007**: Converted NumPy operations to JAX NumPy equivalents

### Code Improvements:
- Replaced @jitclass WealthDynamics with functional NamedTuple approach
- Converted @jit functions to pure JAX functions using jax.lax.scan
- Updated all application code and exercises to use new JAX implementation
- Maintained all original economic functionality while improving performance

## Benefits:
- GPU acceleration support
- Better parallelization with jax.vmap
- Functional programming principles (no mutation)
- Deterministic random number generation
- Advanced JAX optimizations

## Testing:
All existing functionality preserved - the lecture should run identically but with potential for significant performance improvements on GPU hardware.

### CODE REVIEW

**Comment by mmcky** [comment]
## 🔧 Bug Fix: JAX Concretization Errors

Fixed execution errors that occurred during testing of the JAX conversion.

### **Issue Encountered:**
The code was throwing  when executing the  function:


### **Root Cause:**
JAX's  was receiving  as the  parameter, where 
Usage: n [options] [COMMAND] [args]

Commands:

  n                              Display downloaded Node.js versions and install selection
  n latest                       Install the latest Node.js release (downloading if necessary)
  n lts                          Install the latest LTS Node.js release (downloading if necessary)
  n <version>                    Install Node.js <version> (downloading if necessary)
  n install <version>            Install Node.js <version> (downloading if necessary)
  n run <version> [args ...]     Execute downloaded Node.js <version> with [args ...]
  n which <version>              Output path for downloaded node <version>
  n exec <vers> <cmd> [args...]  Execute command with modified PATH, so downloaded node <version> and npm first
  n rm <version ...>             Remove the given downloaded version(s)
  n prune                        Remove all downloaded versions except the installed version
  n --latest                     Output the latest Node.js version available
  n --lts                        Output the latest LTS Node.js version available
  n ls                           Output downloaded versions
  n ls-remote [version]          Output matching versions available for download
  n uninstall                    Remove the installed Node.js
  n download <version>           Download Node.js <version> into cache

Options:

  -V, --version         Output version of n
  -h, --help            Display help information
  -p, --preserve        Preserve npm and npx during install of Node.js
  -q, --quiet           Disable curl output. Disable log messages processing "auto" and "engine" labels.
  -d, --download        Download if necessary. Used with run/exec/which.
  --cleanup             Remove cached version after install
  -a, --arch            Override system architecture
  --offline             Resolve target version against cached downloads instead of internet lookup
  --all                 ls-remote displays all matches instead of last 20
  --insecure            Turn off certificate checking for https requests (may be needed from behind a proxy server)
  --use-xz/--no-use-xz  Override automatic detection of xz support and enable/disable use of xz compressed node downloads.

Aliases:

  install: i
  latest: current
  ls: list
  lsr: ls-remote
  lts: stable
  rm: -
  run: use, as
  which: bin

Versions:

  Numeric version numbers can be complete or incomplete, with an optional leading 'v'.
  Versions can also be specified by label, or codename,
  and other downloadable releases by <remote-folder>/<version>

    4.9.1, 8, v6.1    Numeric versions
    lts               Newest Long Term Support official release
    latest, current   Newest official release
    auto              Read version from file: .n-node-version, .node-version, .nvmrc, or package.json
    engine            Read version from package.json
    boron, carbon     Codenames for release streams
    lts_latest        Node.js support aliases

    and nightly, rc/10 et al was a function argument. When JAX tried to JIT compile the function, it couldn't determine the concrete array size at compile time.

### **Solutions Applied:**

#### ✅ **Fixed  function:**
- **Before:** 
- **After:** 
- Added  to JIT compilation to treat 
Usage: n [options] [COMMAND] [args]

Commands:

  n                              Display downloaded Node.js versions and install selection
  n latest                       Install the latest Node.js release (downloading if necessary)
  n lts                          Install the latest LTS Node.js release (downloading if necessary)
  n <version>                    Install Node.js <version> (downloading if necessary)
  n install <version>            Install Node.js <version> (downloading if necessary)
  n run <version> [args ...]     Execute downloaded Node.js <version> with [args ...]
  n which <version>              Output path for downloaded node <version>
  n exec <vers> <cmd> [args...]  Execute command with modified PATH, so downloaded node <version> and npm first
  n rm <version ...>             Remove the given downloaded version(s)
  n prune                        Remove all downloaded versions except the installed version
  n --latest                     Output the latest Node.js version available
  n --lts                        Output the latest LTS Node.js version available
  n ls                           Output downloaded versions
  n ls-remote [version]          Output matching versions available for download
  n uninstall                    Remove the installed Node.js
  n download <version>           Download Node.js <version> into cache

Options:

  -V, --version         Output version of n
  -h, --help            Display help information
  -p, --preserve        Preserve npm and npx during install of Node.js
  -q, --quiet           Disable curl output. Disable log messages processing "auto" and "engine" labels.
  -d, --download        Download if necessary. Used with run/exec/which.
  --cleanup             Remove cached version after install
  -a, --arch            Override system architecture
  --offline             Resolve target version against cached downloads instead of internet lookup
  --all                 ls-remote displays all matches instead of last 20
  --insecure            Turn off certificate checking for https requests (may be needed from behind a proxy server)
  --use-xz/--no-use-xz  Override automatic detection of xz support and enable/disable use of xz compressed node downloads.

Aliases:

  install: i
  latest: current
  ls: list
  lsr: ls-remote
  lts: stable
  rm: -
  run: use, as
  which: bin

Versions:

  Numeric version numbers can be complete or incomplete, with an optional leading 'v'.
  Versions can also be specified by label, or codename,
  and other downloadable releases by <remote-folder>/<version>

    4.9.1, 8, v6.1    Numeric versions
    lts               Newest Long Term Support official release
    latest, current   Newest official release
    auto              Read version from file: .n-node-version, .node-version, .nvmrc, or package.json
    engine            Read version from package.json
    boron, carbon     Codenames for release streams
    lts_latest        Node.js support aliases

    and nightly, rc/10 et al as static argument

#### ✅ **Fixed  function:**  
- **Before:** 
- **After:** 

### **Technical Details:**
-  has two modes:
  - With  array: iterates over array elements  
  - With  parameter: runs for fixed number of iterations
- Using  avoids concretization issues since JAX knows loop count at compile time
-  tells JAX to compile separate function versions for different parameter values

### **Result:**
✅ All JAX functions now execute without concretization errors  
✅ Code maintains identical functionality with proper JAX compilation  
✅ Ready for performance testing on GPU hardware

**Comment by mmcky** [comment]
## 🔧 Bug Fix: JAX Concretization Errors

Fixed execution errors that occurred during testing of the JAX conversion.

### Issue Encountered:
The code was throwing ConcretizationTypeError when executing the wealth_time_series function due to JAX compilation issues with dynamic array sizes.

### Root Cause:
JAX jax.lax.scan was receiving jnp.arange with dynamic size as the xs parameter. When JAX tried to JIT compile the function, it could not determine the concrete array size at compile time.

### Solutions Applied:

#### ✅ Fixed wealth_time_series function:
- Replaced jnp.arange(n-1) with None, length=n-1 in jax.lax.scan calls
- Added static_argnums=(2,) to JIT compilation to treat n as static argument

#### ✅ Fixed update_cross_section function:  
- Similarly replaced jnp.arange(shift_length) with None, length=shift_length
- This avoids the concretization issue

### Technical Details:
- jax.lax.scan can work with either xs array or length parameter
- Using length avoids concretization issues since JAX knows loop count at compile time
- static_argnums tells JAX to compile separate function versions for different parameter values

### Result:
✅ All JAX functions now execute without concretization errors  
✅ Code maintains identical functionality with proper JAX compilation  
✅ Ready for performance testing on GPU hardware
